tempesta_fw/gfsm.c

```diff
@@ -236,7 +240,7 @@ tfw_gfsm_move(TfwGState *st, unsigned short state, const TfwFsmData *data)
 		 * rather than fixed priority levels to avoid spinning in vain.
 		 */
 		if (!(hooks[p] & mask))
-			return TFW_PASS;
+			goto done;
 
 		/* Switch context to other FSM. */
 		fsm = tfw_gfsm_switch(st, state, p);
```

Comment:
It seems that part of the problem is remain in `tfw_gfsm_switch()`:
https://github.com/tempesta-tech/tempesta/blob/97184f60a78b90ed0703b8c4c4edcefd8290c794/tempesta_fw/gfsm.c#L145
if we have more than one hooks registered for certain FSM state, we get invalid behavior in both cases - whether the previous hook gained `TFW_GFSM_STATE_LAST` or not: `FSM(st)` will access to the FSM of previous hook (or `-1`).
Maybe
```
int fsm_curr = state >> TFW_GFSM_FSM_SHIFT;
int fsm_next = fsm_hooks[fsm_curr][shift].fsm_id;
```
will be more appropriate in this place.

Reply:
Good catch! Fixed.

Reply:
Yeah, really good catch - it takes time to realize the issue.